File: v1/src/main/java/com/api/v1/controllers/CustomerRegistrationController.java
package com.api.v1.controllers;

import com.api.v1.dtos.customer.CustomerRegistrationRequestDto;
import com.api.v1.dtos.customer.CustomerResponseDto;
import com.api.v1.services.customer.CustomerRegistrationService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Mono;

@RestController
@RequestMapping("api/v1/customers")
public class CustomerRegistrationController {

    @Autowired
    private CustomerRegistrationService service;

    @PostMapping
    @ResponseStatus(value = HttpStatus.CREATED)
    public Mono<CustomerResponseDto> register(@Valid @RequestBody CustomerRegistrationRequestDto request) {
        return service.register(request);
    }

}

File: v2/src/main/kotlin/com/api/v2/CustomerRegistrationController.kt
package com.api.v2

import jakarta.validation.Valid
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController
import reactor.core.publisher.Mono

@RestController
@RequestMapping("api/v1/customers")
class CustomerRegistrationController {

    @Autowired
    lateinit var service: CustomerRegistrationService

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    suspend fun register(@RequestBody requestDto: @Valid CustomerRegistrationRequestDto): Mono<CustomerResponseDto> {
        return service.register(requestDto)
    }

}